webapi/src/main/scala/org/knora/webapi/slice/ontology/domain/service/CardinalityService.scala

```diff
@@ -0,0 +1,195 @@
+import org.knora.webapi.queries.sparql._
+import org.knora.webapi.slice.ontology.domain.model.Cardinality
+import org.knora.webapi.slice.ontology.domain.service.ChangeCardinalityCheckResult.CanReplaceCardinalityCheckResult
+import org.knora.webapi.slice.ontology.domain.service.ChangeCardinalityCheckResult.CanReplaceCardinalityCheckResult.CanReplaceCardinalityCheckResult
+import org.knora.webapi.slice.ontology.domain.service.ChangeCardinalityCheckResult.CanReplaceCardinalityCheckResult.CanReplaceCheckSuccess
+import org.knora.webapi.slice.ontology.domain.service.ChangeCardinalityCheckResult.CanReplaceCardinalityCheckResult.IsInUseCheckFailure
+import org.knora.webapi.slice.ontology.domain.service.ChangeCardinalityCheckResult.CanSetCardinalityCheckResult
+import org.knora.webapi.slice.ontology.domain.service.ChangeCardinalityCheckResult.CanSetCardinalityCheckResult.CanSetCardinalityCheckResult
+import org.knora.webapi.slice.resourceinfo.domain.InternalIri
+import org.knora.webapi.slice.resourceinfo.domain.IriConverter
+import org.knora.webapi.store.triplestore.api.TriplestoreService
+
+@accessible
+trait CardinalityService {
+
+  /**
+   * Check if a specific cardinality may be set on a property in the context of a class.
+   *
+   * Setting a wider cardinality on a sub class is not possible if for the same property a stricter cardinality already exists in one of its super classes.
+   *
+   * @param classIri class to check
+   * @param propertyIri property to check
+   * @param newCardinality the newly desired cardinality
+   * @return
+   *    '''success''' a [[CanSetCardinalityCheckResult]] indicating whether a class's cardinalities can be set.
+   *
+   *    '''error''' a [[Throwable]] indicating that something went wrong,
+   */
+  def canSetCardinality(
+    classIri: InternalIri,
+    propertyIri: InternalIri,
+    newCardinality: Cardinality
+  ): Task[CanSetCardinalityCheckResult]
+
+  /**
+   * Check if on a specific class any cardinality may be replaced.
+   *
+   * Replacing an existing cardinality is only possible if the class is not in use yet.
+   *
+   * @param classIri class to check
+   * @return
+   *    '''success''' a [[CanReplaceCardinalityCheckResult]] indicating whether a class's cardinalities can be replaced.
+   *
+   *    '''error''' a [[Throwable]] indicating that something went wrong.
+   */
+  def canReplaceCardinality(classIri: InternalIri): Task[CanReplaceCardinalityCheckResult]
+
+  /**
+   * Check if a property entity is used in resource instances. Returns `true` if
+   * it is used, and `false` if it is not used.
+   *
+   * @param classIri the IRI of the class that is being checked for usage.
+   * @param propertyIri the IRI of the entity that is being checked for usage.
+   *
+   * @return a [[Boolean]] denoting if the property entity is used.
+   */
+  def isPropertyUsedInResources(classIri: InternalIri, propertyIri: InternalIri): Task[Boolean]
+}
+
+object ChangeCardinalityCheckResult {
```

Comment:
Above it's "update", here it is "change". I would be consistent. And, again as above, I think replace and change mean the same thing, or not?

Reply:
"update" or "change" are meant to be a umbrella term for the two different checks:

1. _replace_ check is on a single `classIri`no other entities
2. _set_ checks is on a given set of `classIri`, `propertyIri` and `cardinality`

I have renamed the `RestCardinalityService.canChangeCardinality` method cff8aff2227f6943f3a8c8ce20a2f23207425842.